### PR TITLE
Required Components

### PR DESCRIPTION
**Pull request by cart**
## Introduction

This is the first step in my [Next Generation Scene / UI Proposal](https://github.com/bevyengine/bevy/discussions/14437).

Fixes https://github.com/bevyengine/bevy/issues/7272 #14800.

Bevy's current Bundles as the "unit of construction" hamstring the UI user experience and have been a pain point in the Bevy ecosystem generally when composing scenes:

* They are an additional _object defining_ concept, which must be learned separately from components. Notably, Bundles _are not present at runtime_, which is confusing and limiting.
* They can completely erase the _defining component_ during Bundle init. For example, `ButtonBundle { style: Style::default(), ..default() }` _makes no mention_ of the `Button` component symbol, which is what makes the Entity a "button"!
* They are not capable of representing "dependency inheritance" without completely non-viable / ergonomically crushing nested bundles. This limitation is especially painful in UI scenarios, but it applies to everything across the board.
* They introduce a bunch of additional nesting when defining scenes, making them ugly to look at
* They introduce component name "stutter": `SomeBundle { component_name: ComponentName::new() }`
* They require copious sprinklings of `..default()` when spawning them in Rust code, due to the additional layer of nesting

**Required Components** solve this by allowing you to define which components a given component needs, and how to construct those components when they aren't explicitly provided.

This is what a `ButtonBundle` looks like with Bundles (the current approach):

```rust
#[derive(Component, Default)]
struct Button;

#[derive(Bundle, Default)]
struct ButtonBundle {
    pub button: Button,
    pub node: Node,
    pub style: Style,
    pub interaction: Interaction,
    pub focus_policy: FocusPolicy,
    pub border_color: BorderColor,
    pub border_radius: BorderRadius,
    pub image: UiImage,
    pub transform: Transform,
    pub global_transform: GlobalTransform,
    pub visibility: Visibility,
    pub inherited_visibility: InheritedVisibility,
    pub view_visibility: ViewVisibility,
    pub z_index: ZIndex,
}

commands.spawn(ButtonBundle {
    style: Style {
        width: Val::Px(100.0),
        height: Val::Px(50.0),
        ..default()
    },
    focus_policy: FocusPolicy::Block,
    ..default()
})
```

And this is what it looks like with Required Components:

```rust
#[derive(Component)]
#[require(Node, UiImage)]
struct Button;

commands.spawn((
    Button,
    Style { 
        width: Val::Px(100.0),
        height: Val::Px(50.0),
        ..default()
    },
    FocusPolicy::Block,
));
```

With Required Components, we mention only the most relevant components. Every component required by `Node` (ex: `Style`, `FocusPolicy`, etc)  is automatically brought in!

### Efficiency

1. At insertion/spawn time, Required Components (including recursive required components) are initialized and inserted _as if they were manually inserted alongside the given components_. This means that this is maximally efficient: there are no archetype or table moves.
2. Required components are only initialized and inserted if they were not manually provided by the developer. For the code example in the previous section, because `Style` and `FocusPolicy` are inserted manually, they _will not_ be initialized and inserted as part of the required components system. Efficient!
3. The "missing required components _and_ constructors needed for an insertion" are cached in the "archetype graph edge", meaning they aren't computed per-insertion. When a component is inserted, the "missing required components" list is iterated (and that graph edge (AddBundle) is actually already looked up for us during insertion, because we need that for "normal" insert logic too).

### IDE Integration

The `#[require(SomeComponent)]` macro has been written in such a way that Rust Analyzer can provide type-inspection-on-hover and `F12` / go-to-definition for required components.

### Custom Constructors

The `require` syntax expects a `Default` constructor by default, but it can be overridden with a custom constructor:

```rust
#[derive(Component)]
#[require(
    Node,
    Style(button_style),
    UiImage
)]
struct Button;

fn button_style() -> Style {
    Style {
        width: Val::Px(100.0),
        ..default()
    }
}
```

### Multiple Inheritance

You may have noticed by now that this behaves a bit like "multiple inheritance". One of the problems that this presents is that it is possible to have duplicate requires for a given type at different levels of the inheritance tree:

```rust
#[derive(Component)
struct X(usize);

#[derive(Component)]
#[require(X(x1))
struct Y;

fn x1() -> X {
    X(1)
}

#[derive(Component)]
#[require(
    Y,
    X(x2),
)]
struct Z;

fn x2() -> X {
    X(2)
}

// What version of X is inserted for Z?
commands.spawn(Z);
```

This is allowed (and encouraged), although this doesn't appear to occur much in practice. First: only one version of `X` is initialized and inserted for `Z`. In the case above, I think we can all probably agree that it makes the most sense to use the `x2` constructor for `X`, because `Y`'s `x1` constructor exists "beneath" `Z` in the inheritance hierarchy; `Z`'s constructor is "more specific".

The algorithm is simple and predictable:

1. Use all of the constructors (including default constructors) directly defined in the spawned component's require list
2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a depth Depth First Search). When a constructor is found, it will only be used if one has not already been found.

From a user perspective, just think about this as the following:

1. Specifying a required component constructor for `Foo` directly on a spawned component `Bar` will result in that constructor being used (and overriding existing constructors lower in the inheritance tree). This is the classic "inheritance override" behavior people expect.
2. For cases where "multiple inheritance" results in constructor clashes, Components should be listed in "importance order". List a component earlier in the requirement list to initialize its inheritance tree earlier.

Required Components _does_ generally result in a model where component values are decoupled from each other at construction time. Notably, some existing Bundle patterns use bundle constructors to initialize multiple components with shared state. I think (in general) moving away from this is necessary:

1. It allows Required Components (and the Scene system more generally) to operate according to simple rules
2. The "do arbitrary init value sharing in Bundle constructors" approach _already_ causes data consistency problems, and those problems would be exacerbated in the context of a Scene/UI system. For cases where shared state is truly necessary, I think we are better served by observers / hooks.
3. If a situation _truly_ needs shared state constructors (which should be rare / generally discouraged), Bundles are still there if they are needed.

## Next Steps

* **Require Construct-ed Components**: I have already implemented this (as defined in the [Next Generation Scene / UI Proposal](https://github.com/bevyengine/bevy/discussions/14437). However I've removed `Construct` support from this PR, as that has not landed yet. Adding this back in requires relatively minimal changes to the current impl, and can be done as part of a future Construct pr.
* **Port Built-in Bundles to Required Components**: This isn't something we should do right away. It will require rethinking our public interfaces, which IMO should be done holistically after the rest of Next Generation Scene / UI lands. I think we should merge this PR first and let people experiment _inside their own code with their own Components_ while we wait for the rest of the new scene system to land.
* **_Consider_ Automatic Required Component Removal**: We should evaluate _if_ automatic Required Component removal should be done. Ex: if all components that explicitly require a component are removed, automatically remove that component. This issue has been explicitly deferred in this PR, as I consider the insertion behavior to be desirable on its own (and viable on its own). I am also doubtful that we can find a design that has behavior we actually want. Aka: can we _really_ distinguish between a component that is "only there because it was automatically inserted" and "a component that was necessary / should be kept".  See my [discussion response here](https://github.com/bevyengine/bevy/discussions/14437#discussioncomment-10268668) for more details.

### CODE REVIEW

File: crates/bevy_ecs/src/bundle.rs [hunk 1]
```diff
@@ -475,6 +508,74 @@ impl BundleInfo {
...
+    /// Internal method to initialize a required component from an [`OwningPtr`]. This should ultimately be called
+    /// in the context of [`BundleInfo::write_components`], via [`RequiredComponent::initialize`].
+    ///
+    /// # Safety
+    ///
+    /// `component_ptr` must point to a required component value that matches the given `component_id`. The `storage_type` must match
+    /// the type associated with `component_id`. The `entity` and `table_row` must correspond to an entity with an uninitialized
+    /// component matching `component_id`.
+    ///
+    /// This _should not_ be used outside of [`BundleInfo::write_components`].
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/bundle.rs:534]:
```suggestion
    /// This method _should not_ be called outside of [`BundleInfo::write_components`].
```

File: crates/bevy_ecs/src/bundle.rs [hunk 2]
```diff
@@ -174,6 +177,13 @@ pub unsafe trait Bundle: DynamicBundle + Send + Sync + 'static {
         // Ensure that the `OwningPtr` is used correctly
         F: for<'a> FnMut(&'a mut T) -> OwningPtr<'a>,
         Self: Sized;
+
+    /// Registers components that are required by the components in this [`Bundle`].
+    fn register_required_components(
+        _components: &mut Components,
+        _storages: &mut Storages,
+        _required_components: &mut RequiredComponents,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/bundle.rs:185]: The fact that this reference needs to be mutable is surprising to me.

File: crates/bevy_ecs/src/component.rs [hunk 3]
```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
         }
     }
 }
+#[cfg(feature = "track_change_detection")]
+type RequiredComponentConstructor =
+    dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity, &'static Location<'static>);
+#[cfg(not(feature = "track_change_detection"))]
+type RequiredComponentConstructor = dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity);
+
+/// A context-less constructor that can be called to construct a new instance of a component.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:1154]: These docs are quite odd to me; they don't actual mention / tie into the notion of what a required component *is*, and instead focus purely on the mechanics.

File: crates/bevy_ecs/src/component.rs [hunk 4]
```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
...
+            change_tick,
+            table_row,
+            entity,
+            #[cfg(feature = "track_change_detection")]
+            caller,
+        );
+    }
+}
+
+/// The collection of metadata for components that are required for a given component.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:1195]: This either needs a doc test demonstrating how to use the macro to register required components or a link to the docs where this *is* demonstrated.

File: crates/bevy_ecs/src/component.rs [hunk 5]
```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
...
+        // matches the storage type of the component. It only uses the given `table_row` or `Entity` to
+        // initialize the storage corresponding to the given entity.
+        unsafe { self.register_dynamic(component_id, erased) };
+    }
+
+    /// Removes components that are explicitly provided in a given [`Bundle`]. These components should
+    /// be logically treated as normal components, not "required components".
+    ///
+    /// [`Bundle`]: crate::bundle::Bundle
+    pub fn remove_bundle_components(&mut self, components: &[ComponentId]) {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:1270]: I think `remove_explicit_components` or something would be more clear here.

File: crates/bevy_ecs/src/lib.rs [hunk 6]
```diff
@@ -1793,6 +1793,78 @@ mod tests {
...
+
+        impl Default for Y {
+            fn default() -> Self {
+                Self {
+                    value: "hello".to_string(),
+                }
+            }
+        }
+
+        fn new_z() -> Z {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/lib.rs:1819]: This would be clearer as `construct_z` or something: `new_z` read as a setter to me.

**Review by alice-i-cecile — COMMENTED** [review]
I really like the end result, but have a few nits on internal clarity. I'd also really like to ship this with good docs demonstrating the various features, probably on the `Component` trait.

File: crates/bevy_ecs/src/bundle.rs [hunk 7]
```diff
@@ -1128,12 +1259,16 @@ impl Bundles {
             let mut component_ids = Vec::new();
             T::component_ids(components, storages, &mut |id| component_ids.push(id));
             let id = BundleId(bundle_infos.len());
+            let mut required_components = RequiredComponents::default();
+            T::register_required_components(components, storages, &mut required_components);
+            required_components.remove_bundle_components(&component_ids);
+            let required_components = required_components.0.drain().map(|(_, v)| v).collect();
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/bundle.rs:1265]: Draining mapping and recollecting feels wasteful. Why can't we just pass in required_components.values()?

File: crates/bevy_ecs/src/archetype.rs [hunk 8]
```diff
@@ -123,6 +123,8 @@ pub(crate) struct AddBundle {
     /// For each component iterated in the same order as the source [`Bundle`](crate::bundle::Bundle),
     /// indicate if the component is newly added to the target archetype or if it already existed
     pub bundle_status: Vec<ComponentStatus>,
+    /// Required components that must be initialized immediately when adding this Bundle.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/archetype.rs:126]:
```suggestion
    /// The set of additional required components that must be initialized immediately when adding this Bundle.
    ///
    /// The initial values are determined based on the provided constructor, falling back to the `Default` trait if none is given.
```

File: crates/bevy_ecs/src/component.rs [hunk 9]
```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
...
+            table_row,
+            entity,
+            #[cfg(feature = "track_change_detection")]
+            caller,
+        );
+    }
+}
+
+/// The collection of metadata for components that are required for a given component.
+#[derive(Default)]
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:1196]:
```suggestion
#[derive(Default, Debug)]
```

File: crates/bevy_ecs/src/component.rs [hunk 10]
```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
...
     }
 }
+#[cfg(feature = "track_change_detection")]
+type RequiredComponentConstructor =
+    dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity, &'static Location<'static>);
+#[cfg(not(feature = "track_change_detection"))]
+type RequiredComponentConstructor = dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity);
+
+/// A context-less constructor that can be called to construct a new instance of a component.
+#[derive(Clone)]
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:1155]:
```suggestion
#[derive(Clone, Debug)]
```

**Comment by alice-i-cecile** [comment]
One more question: it looks like the fallback constructor is `Default`. Why not make it `FromWorld` for both consistency and more expressiveness?

**Comment by james-j-obrien** [comment]
Seems like this doesn't work with hooks/observers as the required components aren't in the added array in `BundleInfo`.

Here's a test that fails:
```rust
#[test]
fn required_components_hooks() {
    #[derive(Component)]
    #[require(Y)]
    struct X;

    #[derive(Component, Default)]
    struct Y;

    #[derive(Resource)]
    struct R(usize);

    let mut world = World::new();
    world.insert_resource(R(0));
    world
        .register_component_hooks::<Y>()
        .on_add(|mut world, _, _| world.resource_mut::<R>().0 += 1);

    // Spawn entity and ensure Y was added
    assert!(world.spawn(X).contains::<Y>());

    assert_eq!(world.resource::<R>().0, 1);
}
```

**Comment by cart** [comment]
> One more question: it looks like the fallback constructor is Default. Why not make it FromWorld for both consistency and more expressiveness?

Given that the constructors are called mid-insertion, this is dangerous. FromWorld could in theory access data that shouldn't be accessed during an insert. Ex: try to read components before they are initialized.

In theory I suspect we would be able to do this via careful ordering of operations, but it would also require some serious restructuring of entity initialization (and how we break up / borrow components from world), especially if we don't want to redundantly allocate outside of the "real" storage.

Worth considering, but imo as a separate PR, and with _much_ more meticulous soundness analysis.

**Comment by cart** [comment]
> Seems like this doesn't work with hooks/observers as the required components aren't in the added array in BundleInfo (which is the current source of truth).

Hmm yeah I'll look into this.

File: crates/bevy_ecs/src/bundle.rs (hunk 2 above)
Comment (cart) [reply crates/bevy_ecs/src/bundle.rs:185]: `_components` and `_storages` are `mut` because we need to be able to initialize components that haven't been initialized yet (to retrieve the ComponentId).

`_required_components`  is `mut` because it is the collection of components, which we are adding to in this function. This is where the "registration" happens.

File: crates/bevy_ecs/src/bundle.rs (hunk 7 above)
Comment (cart) [reply crates/bevy_ecs/src/bundle.rs:1265]: BundleInfo needs an owned collection of RequiredComponent, so `required_components.values()` doesn't work here because it is a borrowing iterator.

In theory we could just store the HashMap directly instead of converting it to a Vec, but:

1. With the addition of `require(construct(X))`, the map will have entries that could either be Construct or Immediate. These will then need to be split out into separate lists for BundleInfo.
2. The `values()` HashMap iterator is slower than iterating a Vec. Given that iterating the list will be in the much hotter path of "work done for every entity", doing the work here (once) to convert that to a Vec will almost certainly pay dividends in the majority of cases.

**Comment by tychedelia** [comment]
Confirmed that go to definition works in rust-analyzer:

image

Unfortunately, it does not work in RustRover :(
image

Not sure if there's prior art here that's already on JetBrains radar, but it might be worth filing an issue if not.

**Comment by TotalKrill** [comment]
I have a question regarding the ergnomics, using the Button example provided. 

How in the following example would I find out that I can/need to spawn  Style to change the look of my spawned Button, neither the Docs or the Language server would help me in this example.
Traversing  the tree from Button->Node->Wherever, and doing the same for UiImage Feels like I would get lost quite quickly.

```
#[derive(Component)]
#[require(Node, UiImage)]
struct Button;

commands.spawn((
    Button,
    Style { 
        width: Val::Px(100.0),
        height: Val::Px(50.0),
        ..default()
    },
    FocusPolicy::Block,
```

File: crates/bevy_ecs/macros/src/component.rs [hunk 11]
```diff
@@ -66,12 +73,40 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
...
         impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
             const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;
+            fn register_required_components(
+                components: &mut #bevy_ecs_path::component::Components,
+                storages: &mut #bevy_ecs_path::storage::Storages,
+                required_components: &mut #bevy_ecs_path::component::RequiredComponents
+            ) {
+                #(#register_required)*
+                #(#register_recursive_requires)*
+            }
```
Comment (SkiFire13) [inline crates/bevy_ecs/macros/src/component.rs:109]: Could this document the reasoning for ordering `register_required` and `register_recursive_requires`? It seems critical for ensuring that the correct constructor is used for components that are required by more than one other component.

File: crates/bevy_ecs/macros/src/component.rs [hunk 12]
```diff
@@ -116,42 +159,64 @@ fn parse_component_attr(ast: &DeriveInput) -> Result<Attrs> {
...
+                    Ok(())
+                } else {
+                    Err(nested.error("Unsupported attribute"))
+                }
+            })?;
+        } else if attr.path().is_ident(REQUIRE) {
+            let punctuated =
+                attr.parse_args_with(Punctuated::<Require, Comma>::parse_terminated)?;
+            attrs.requires = Some(punctuated);
+        }
```
Comment (SkiFire13) [inline crates/bevy_ecs/macros/src/component.rs:199]: It seems this will silently ignore repeated `requires` attributes. IMO this should instead either error on multiple `requires` or merge them into one.

File: crates/bevy_ecs/macros/src/component.rs [hunk 13]
```diff
@@ -66,12 +73,40 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
...
+                register_required.push(quote! {
+                    required_components.register(components, storages, || { let x: #ident = #func().into(); x });
+                });
+            } else {
+                register_required.push(quote! {
+                    required_components.register(components, storages, <#ident as Default>::default);
+                });
+            }
+        }
+    }
```
Comment (SkiFire13) [inline crates/bevy_ecs/macros/src/component.rs:95]: It seems this will silently succeeds if `requires` contains the same component multiple times. Wouldn't it be better if it raised an error/warning?

File: crates/bevy_ecs/src/bundle.rs (hunk 7 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/bundle.rs:1265]: Performance-wise this seems fine to me since it will be called only once per `Bundle`.

The only thing weird seems to be the `.drain()`. Why not `.into_iter()`?

File: crates/bevy_ecs/macros/src/component.rs (hunk 12 above)
Comment (BD103) [reply crates/bevy_ecs/macros/src/component.rs:199]: I'm in favor of them being merged, since that would make life easier for macros that automatically generate the `#[requires(...)]` attribute.

File: crates/bevy_ecs/src/bundle.rs (hunk 7 above)
Comment (BD103) [reply crates/bevy_ecs/src/bundle.rs:1265]: `into_iter()` seems like the better choice in this example. If you look at the documentation for `HashMap::drain()`, it says:

> "Clears the map, returning all key-value pairs as an iterator. Keeps the allocated memory for reuse." ([Source](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.drain))

It looks like `drain()` is best when you intend on re-using the `HashMap`, but in this case `required_components` gets shadowed then dropped at the end of this function.

```suggestion
            let required_components = required_components.0.into_iter().map(|(_, v)| v).collect();
```

File: crates/bevy_ecs/src/bundle.rs [hunk 14]
```diff
@@ -1230,7 +1365,7 @@ fn initialize_dynamic_bundle(
     let id = BundleId(bundle_infos.len());
     let bundle_info =
         // SAFETY: `component_ids` are valid as they were just checked
-        unsafe { BundleInfo::new("<dynamic bundle>", components, component_ids, id) };
+        unsafe { BundleInfo::new("<dynamic bundle>", components, component_ids, Vec::new(), id) };
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/bundle.rs:1368]: This `Vec::new()` seems wrong, as it means that dynamically inserting a component won't insert its required components.

**Comment by Zeenobit** [comment]
Thank you for this! 🚀 

Does this implementation allow users to query the required components of a component at runtime?

Context:
In [Moonshine Kind](https://github.com/Zeenobit/moonshine_kind/blob/20ecc2069bbece8419b85d3a4a0372463596a7d6/src/lib.rs#L78), I use a `CastInto<T>` trait which has to be implemented manually to allow "upcasting" safely between related components (i.e. a `Button` should always have a `Node`, so an `Instance<Button>` is safely convertible to `Instance<Node>`).

If we can query the required components, safe casting may be doable automatically.

**Comment by DasLixou** [comment]
@Zeenobit still not safe, user can remove required component (e.g. `Node`) but the root component (e.g. `Button`) can stay

**Comment by iiYese** [comment]
I'll do a pass of the code later but 2 things about the API design:
- Would this cause problems/be incompatible with private `Component`s? Crates will have private `Component`s they might expose `Bundle`s & `QueryData`s for.
- What if I want a constructor that needs input? This scenario is better covered by relations & inheritance but just to illustrate the problem I might want a `EntityRef` so I could get a `Style` component from the parent.

**Comment by Zeenobit** [comment]
> @Zeenobit still not safe, user can remove required component (e.g. `Node`) but the root component (e.g. `Button`) can stay

That's a known (and in my opinion, safely ignorable) issue: I just don't wouldn't use requirements and/or kind semantics on components that I expect to be removed post initialization.

**Comment by killercup** [comment]
Added https://github.com/bevyengine/bevy/issues/14800 with a suggestion to also generate doc comments from this. (Didn't want to add this here so we can land this quicker.)

File: crates/bevy_ecs/src/bundle.rs (hunk 14 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/bundle.rs:1368]: And here's a test that fails due to this:

```rs
#[derive(Component)]
#[require(Y)]
struct X;

#[derive(Component, Default)]
struct Y;

let mut world = World::new();
let x_id = world.init_component::<X>();

let mut e = world.spawn_empty();

bevy_ptr::OwningPtr::make(X, |ptr| unsafe {
    e.insert_by_id(x_id, ptr);
});

assert!(e.contains::<Y>());
```

File: crates/bevy_ecs/src/lib.rs (hunk 6 above)
Comment (cart) [reply crates/bevy_ecs/src/lib.rs:1819]: I would prefer to not use "construct" here, given that that term will mean something specific soon, and it will relate to this code.

Comment (cart) [reply crates/bevy_ecs/src/lib.rs:1819]: I don't see how `new_z` would read as a setter. I can't think of any cases where I've seen that convention. I've seen plenty of cases where `new_x` initializes a new X though.

File: crates/bevy_ecs/src/bundle.rs (hunk 7 above)
Comment (cart) [reply crates/bevy_ecs/src/bundle.rs]: I've swapped to into_iter. That does make more sense.

File: crates/bevy_ecs/macros/src/component.rs (hunk 12 above)
Comment (cart) [reply crates/bevy_ecs/macros/src/component.rs:199]: Merging makes sense to me!

File: crates/bevy_ecs/macros/src/component.rs (hunk 13 above)
Comment (cart) [reply crates/bevy_ecs/macros/src/component.rs:95]: This is _slightly_ nontrivial because using just the built-in hashset storage, by default theres no way to tell if a component was contributed by the current required component (redundant + likely invalid) or contributed by a previously visited component "higher" in the tree (totally valid).

We'd need to allocate and populate a second hashset during init (more runtime work), or store the component that contributed the required component in the current hashset, so it can be compared (less runtime work, but adds complexity to the registration API).

The latter strategy probably makes the most sense, provided we're willing to add that to the registration API.

**Comment by cart** [comment]
Just slimmed down the size of `AddBundle` by only storing the Required Component constructor on it (instead of also including the ComponentId). Given that AddBundle will be high-frequency, better to cut down on init and memory costs.

File: crates/bevy_ecs/macros/src/component.rs (hunk 13 above)
Comment (cart) [reply crates/bevy_ecs/macros/src/component.rs:95]: I suppose we could also just compare the identifier paths in the macro and fail statically. That would cut down on runtime overhead, at the cost of not catching cases where two paths point to the same identifier. I think that actually has my preference, given that this is per-bundle work being done. I prefer to minimize overhead as much as possible.

CONST TYPEID WHERE ARE YOU

Comment (SkiFire13) [reply crates/bevy_ecs/macros/src/component.rs:95]: My comment was just about the identifier paths that the macro sees. Checking them is of course not perfect, but at least avoids clearly non-sensical attributes

**Comment by cart** [comment]
A collection of "big" todos for this PR:

1. Support generic requires (this was on my pre-PR todo list but I missed it)
2. Support hooks / observers (currently does not invoke them)
3. Support dynamic components (currently not wired up)

File: crates/bevy_ecs/src/component.rs [hunk 15]
```diff
@@ -93,6 +97,91 @@ use std::{
...
+/// struct B(usize);
+///
+/// #[derive(Component, Default)]
+/// struct C(f32);
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B and C with their Default constructors
+/// world.spawn(A);
+/// ```
+///
```
Comment (iiYese) [inline crates/bevy_ecs/src/component.rs:184]: What happens in the following scenario?
```rs
#[derive(Component, Default)]
#[require(B<usize>)]
struct B<T>(T);

world.spawn(B(1_u8));
```
Does it create `(B(1_u8), B(0_usize))` or does `B(0_usize)` try require itself & do something weird?

**Review by iiYese — COMMENTED** [review]
Reviewed docs, tests & tried to find edge cases.

File: crates/bevy_ecs/src/component.rs [hunk 16]
```diff
@@ -1133,3 +1230,151 @@ impl<T: Component> FromWorld for InitComponentId<T> {
...
+            change_tick,
+            table_row,
+            entity,
+            #[cfg(feature = "track_change_detection")]
+            caller,
+        );
+    }
+}
+
+/// A component that another component must when it is inserted. This contains a context-less constructor that
```
[the remark this answers is not in the document]
Comment (BD103) [reply crates/bevy_ecs/src/component.rs:1279]:
```suggestion
/// A component that another component must have when it is inserted. This contains a context-less constructor that
```

**Comment by cart** [comment]
Just fixed hooks/observers, which included a refactor of BundleInfo + AddBundle to improve clarity about the scope of the components being accessed. Also added more tests.

File: crates/bevy_ecs/src/component.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (cart) [reply crates/bevy_ecs/src/component.rs:184]: Cycles result in a stack overflow at insert time.

Comment (cart) [reply crates/bevy_ecs/src/component.rs:184]: This would result in a stack overflow at insert time, as `B<usize>` would also require `B<usize>`.

File: crates/bevy_ecs/src/component.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (cart) [reply crates/bevy_ecs/src/component.rs]: This type has been removed!

File: crates/bevy_ecs/src/bundle.rs (hunk 14 above)
Comment (cart) [reply crates/bevy_ecs/src/bundle.rs]: I've reworked the dataflow a bit to make this possible (RequiredComponents are now stored on ComponentInfo). Thanks for the test. It now passes!

File: crates/bevy_ecs/macros/src/component.rs (hunk 13 above)
Comment (cart) [reply crates/bevy_ecs/macros/src/component.rs]: I've added this!

File: crates/bevy_ecs/src/component.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (cart) [reply crates/bevy_ecs/src/component.rs:184]: Added a doc about this

**Comment by cart** [comment]
I've addressed all comments. The big changes since the opening the PR:

* Generics and full type paths in `#[requires()]` are now supported
* Hooks and observers now work with required components
* Requires now work with dynamic bundles / components
* Required Components can now be queried for a given component using `let info = world.components().get_info(id).unwrap()` followed by `info.required_component.iter_ids()`
* The derive macro now generates docs containing a list of the required components (with links to their docs). This is accomplished by adding the doc to the `Component` trait impl. Sadly we can't put the docs _directly_ on the type's docs, but this still works pretty well:
  ```rust
  #[derive(Component)]
  #[require(B, C)]
  pub struct A;
  ```
  image
 
For some reason miri builds are failing because syn is missing some impls with that build config (not a soundness issue). Not quite sure what the deal is there yet. Once I've solved that I think this is good to go.

File: crates/bevy_ecs/macros/src/component.rs [hunk 17]
```diff
@@ -66,12 +75,55 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
...
     let struct_name = &ast.ident;
     let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
 
+    let required_component_docs = attrs.requires.map(|r| {
+        let paths = r
+            .iter()
+            .map(|r| format!("[`{}`]", r.path.to_token_stream()))
+            .collect::<Vec<_>>()
+            .join(", ");
+        let doc = format!("Required Components: {paths}. \n\n A component's Required Components are inserted whenever it is inserted. Note that this will also insert the required components _of_ the required components, recursively.");
```
Comment (torsteingrindvik) [inline crates/bevy_ecs/macros/src/component.rs:107]: Consider a mention that it's depth first too

File: crates/bevy_ecs/src/component.rs [hunk 18]
```diff
@@ -93,6 +97,94 @@ use std::{
...
+/// #[derive(Component, Default)]
+/// #[require(C)]
+/// struct B(usize);
+///
+/// #[derive(Component, Default)]
+/// struct C(f32);
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B and C with their Default constructors
+/// world.spawn(A);
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:182]: Nit: It might be nice to actually toss in an assert here to make this more test-like. More tests is good, but it's also reassuring (and informative) to users.

**Review by alice-i-cecile — APPROVED** [review]
Excellent: great docs, clean implementation, tests for all of the nasty edge cases. I think this is ready to merge.

File: crates/bevy_ecs/src/component.rs (hunk 18 above)
Comment (cart) [reply crates/bevy_ecs/src/component.rs]: Good call! Done.

**Review by NthTensor — APPROVED** [review]
Wow, this is really well done. The docs on this are especially high quality. Thank you for prioritizing this section of the bsn work.

It looks like the notes on multiple inheritance didn't really make it into the docs. Could we copy over the explanation from the PR description?

**Comment by cart** [comment]
> It looks like the notes on multiple inheritance didn't really make it into the docs. Could we copy over the explanation from the PR description?

Done!

**Comment by alice-i-cecile** [comment]
@cart feel free to merge once CI is green: the new docs look good and it looks like you've resolved the remaining concerns for the MVP.

File: crates/bevy_ecs/src/component.rs [hunk 19]
```diff
@@ -93,6 +97,141 @@ use std::{
...
+///
+/// # let mut world = World::default();
+/// // In this case, the x2 constructor is used for X
+/// let id = world.spawn(Z).id();
+/// assert_eq!(2, world.entity(id).get::<X>().unwrap().0);
+/// ```
+///
+/// In general, this shouldn't happen often, but when it does the algorithm is simple and predictable:
+/// 1. Use all of the constructors (including default constructors) directly defined in the spawned component's require list
+/// 2. In the order the requires are defined in #[require()], recursively visit the require list of each of the components in the list (this is a depth Depth First Search). When a constructor is found, it will only be used if one has not already been found.
```
Comment (killercup) [inline crates/bevy_ecs/src/component.rs:229]:
```suggestion
/// 2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a depth Depth First Search). When a constructor is found, it will only be used if one has not already been found.
```

**Review by killercup — COMMENTED** [review]
Added suggestion for what CI complained about

**Comment by Jaso333** [comment]
I've joined the party a bit late here, but I'm wondering how this affects a certain pattern I'm using.

I'm currently using populated bundles as templates for specific types of entities, e.g. I have a resource: `EnemyTemplate(EnemyBundle)` which gets created by a system at some point after the assets have been loaded; it inserts the resource into the world containing the bundle with the correct sprite image and other configuration from the world. This is kind of like my own rudimentary "prefab" pattern. If the idea of "bundles" is deprecated, I don't really know how to achieve this, or at least it isn't clear to me. I understand that bundles are still here to stay, but I'd like to avoid them if I can in favour of whatever the new way is of doing this.

**Comment by NthTensor** [comment]
Bundles themselves are not deprecated, but the existing bundles are being removed from the engine's public api. Required components is actually kind of built on bundles, actually.

In the future, the recommended approach to this will be through BSN. For now, you may have to write some custom bundles for yourself.

**Comment by Jaso333** [comment]
Understood. At least I won't have bundle trees like I currently have, e.g. `EnemyBundle` contains a field for `SpriteBundle`.

**Comment by Zeenobit** [comment]
Are component requirements enforced during scene deserialization?

i.e. if component A requires B, and I load a scene with just an A, would B get added dynamically?

**Comment by alice-i-cecile** [comment]
Yes :)

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1725 if you'd like to help out.
